File: common/src/main/java/com/dreamdisplays/ytdlp/YtVideoInfo.java
package com.dreamdisplays.ytdlp;

import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

// TODO: remove
@NullMarked
public final class YtVideoInfo {

    private final String id;
    private final String title;
    private final @Nullable String uploader;
    private final @Nullable Long durationSec;
    private final @Nullable Long viewCount;
    private final @Nullable Long likeCount;
    private final @Nullable String publishedText;
    private final @Nullable Integer publishedDaysAgo;
    private final @Nullable String thumbnailUrl;

    public YtVideoInfo(
            String id,
            String title,
            @Nullable String uploader,
            @Nullable Long durationSec,
            @Nullable Long viewCount,
            @Nullable String thumbnailUrl
    ) {
        this(id, title, uploader, durationSec, viewCount, null, null, null, thumbnailUrl);
    }

    public YtVideoInfo(
            String id,
            String title,
            @Nullable String uploader,
            @Nullable Long durationSec,
            @Nullable Long viewCount,
            @Nullable Long likeCount,
            @Nullable String publishedText,
            @Nullable Integer publishedDaysAgo,
            @Nullable String thumbnailUrl
    ) {
        this.id = id;
        this.title = title;
        this.uploader = uploader;
        this.durationSec = durationSec;
        this.viewCount = viewCount;
        this.likeCount = likeCount;
        this.publishedText = publishedText;
        this.publishedDaysAgo = publishedDaysAgo;
        this.thumbnailUrl = thumbnailUrl;
    }

    public String getId() {
        return id;
    }

    public String getTitle() {
        return title;
    }

    public @Nullable String getUploader() {
        return uploader;
    }

    public @Nullable Long getDurationSec() {
        return durationSec;
    }

    public @Nullable Long getViewCount() {
        return viewCount;
    }

    public @Nullable Long getLikeCount() {
        return likeCount;
    }

    public @Nullable String getPublishedText() {
        return publishedText;
    }

    public @Nullable Integer getPublishedDaysAgo() {
        return publishedDaysAgo;
    }

    public boolean isRecent(int daysWindow) {
        return publishedDaysAgo != null && publishedDaysAgo >= 0 && publishedDaysAgo <= daysWindow;
    }

    public String getWatchUrl() {
        return "https://www.youtube.com/watch?v=" + id;
    }

    public String getThumbnailUrl() {
        return "https://i.ytimg.com/vi/" + id + "/mqdefault.jpg";
    }

    public String formatDuration() {
        if (durationSec == null || durationSec <= 0) return "";
        long s = durationSec;
        long h = s / 3600;
        long m = (s % 3600) / 60;
        long sec = s % 60;
        if (h > 0) return String.format("%d:%02d:%02d", h, m, sec);
        return String.format("%d:%02d", m, sec);
    }

    public String formatViews() {
        if (viewCount == null || viewCount <= 0) return "";
        if (viewCount >= 1_000_000_000L)
            return String.format("%.1fB views", viewCount / 1_000_000_000.0);
        if (viewCount >= 1_000_000L)
            return String.format("%.1fM views", viewCount / 1_000_000.0);
        if (viewCount >= 1_000L)
            return String.format("%.1fK views", viewCount / 1_000.0);
        return viewCount + " views";
    }

    public String formatLikes() {
        if (likeCount == null || likeCount <= 0) return "";
        if (likeCount >= 1_000_000_000L)
            return String.format("%.1fB", likeCount / 1_000_000_000.0);
        if (likeCount >= 1_000_000L)
            return String.format("%.1fM", likeCount / 1_000_000.0);
        if (likeCount >= 1_000L)
            return String.format("%.1fK", likeCount / 1_000.0);
        return likeCount.toString();
    }
}

File: common/src/main/kotlin/com/dreamdisplays/ytdlp/YtVideoInfo.kt
package com.dreamdisplays.ytdlp

/**
 * Data class representing YouTube video information, as returned by `yt-dlp`. It includes methods to format the
 * duration, view count, and like count.
 */
class YtVideoInfo(
    val id: String,
    val title: String,
    val uploader: String?,
    private val durationSec: Long?,
    private val viewCount: Long?,
    private val likeCount: Long? = null,
    val publishedText: String? = null,
    private val publishedDaysAgo: Int? = null,
) {

    /** Returns true if the video is published within the last [daysWindow] days. */
    fun isRecent(daysWindow: Int): Boolean =
        publishedDaysAgo != null && publishedDaysAgo >= 0 && publishedDaysAgo <= daysWindow

    /** Returns the YouTube watch URL for this video. */
    fun getWatchUrl(): String = "https://www.youtube.com/watch?v=$id"

    /** Returns the YouTube thumbnail URL for this video. */
    fun getThumbnailUrl(): String = "https://i.ytimg.com/vi/$id/mqdefault.jpg"

    /** Returns a formatted string representation of the video duration in HH:MM:SS format. */
    fun formatDuration(): String {
        val s = durationSec ?: return ""
        if (s <= 0) return ""
        val h = s / 3600
        val m = (s % 3600) / 60
        val sec = s % 60
        return if (h > 0) String.format("%d:%02d:%02d", h, m, sec)
        else String.format("%d:%02d", m, sec)
    }

    /** Returns a formatted string representation of the video view count. */
    fun formatViews(): String {
        val v = viewCount ?: return ""
        if (v <= 0) return ""
        return when {
            v >= 1_000_000_000L -> String.format("%.1fB views", v / 1_000_000_000.0)
            v >= 1_000_000L -> String.format("%.1fM views", v / 1_000_000.0)
            v >= 1_000L -> String.format("%.1fK views", v / 1_000.0)
            else -> "$v views"
        }
    }

    /** Returns a formatted string representation of the video like count. */
    fun formatLikes(): String {
        val l = likeCount ?: return ""
        if (l <= 0) return ""
        return when {
            l >= 1_000_000_000L -> String.format("%.1fB", l / 1_000_000_000.0)
            l >= 1_000_000L -> String.format("%.1fM", l / 1_000_000.0)
            l >= 1_000L -> String.format("%.1fK", l / 1_000.0)
            else -> l.toString()
        }
    }
}
